File: crates/solidity/outputs/npm/tests/src/doc-examples/using-queries.ts
import { repoPath } from "../utils/files";
import path from "node:path";
import assert from "node:assert";
import fs from "node:fs/promises";

import { Language } from "@nomicfoundation/slang/language";
import { NonterminalKind } from "@nomicfoundation/slang/kinds";
import { Query, QueryMatchIterator } from "@nomicfoundation/slang/query";
import { NonterminalNode, TerminalNode } from "@nomicfoundation/slang/cst";

async function parseDocInputFile(filePath: string) {
  const inputPath = path.join(repoPath("documentation/public/user-guide/inputs"), filePath);
  const source = await fs.readFile(inputPath, "utf8").then((source) => source.trim());
  const language = new Language("0.8.0");
  return language.parse(NonterminalKind.SourceUnit, source);
}

test("using queries", async () => {
  {
    const parse_output = await parseDocInputFile("using-the-cursor.sol");
    // --8<-- [start:creating-a-query]

    // Any `Cursor` can be used to create a query.
    const cursor = parse_output.createTreeCursor();

    const query = Query.parse("[ContractDefinition]");
    const matches: QueryMatchIterator = cursor.query([query]);
    // --8<-- [end:creating-a-query]
    matches; // Silence the unused warning
  }

  {
    const parseOutput = await parseDocInputFile("using-the-cursor.sol");
    const cursor = parseOutput.createTreeCursor();
    // --8<-- [start:visiting-contracts]
    const found = [];

    const query = Query.parse("@contract [ContractDefinition]");
    const matches = cursor.query([query]);

    let match = null;
    while ((match = matches.next())) {
      const captures = match.captures;
      const cursors = captures["contract"];

      const cursor = cursors?.[0]?.node() as NonterminalNode;

      found.push(cursor.unparse().trim());
    }

    assert.deepStrictEqual(found, ["contract Foo {}", "contract Bar {}", "contract Baz {}"]);
    // --8<-- [end:visiting-contracts]
  }

  {
    const parse_output = await parseDocInputFile("multiple-data-types.sol");
    const cursor = parse_output.createTreeCursor();
    // --8<-- [start:multiple-patterns]
    const names = [];

    const struct_def = Query.parse("[StructDefinition @name [Identifier]]");
    const enum_def = Query.parse("[EnumDefinition @name [Identifier]]");
    const matches = cursor.query([struct_def, enum_def]);

    let match = null;
    while ((match = matches.next())) {
      const index = match.queryNumber;
      const captures = match.captures;
      const cursors = captures["name"];

      const cursor = cursors?.[0];

      names.push([index, (cursor?.node() as TerminalNode).text]);
    }

    assert.deepStrictEqual(names, [
      [0, "Foo"],
      [1, "Bar"],
      [0, "Baz"],
      [1, "Qux"],
    ]);
    // --8<-- [end:multiple-patterns]
  }

  {
    const parse_output = await parseDocInputFile("typed-tuple.sol");
    const cursor = parse_output.createTreeCursor();
    // --8<-- [start:matching-on-label]

    const names = [];

    const query = Query.parse("[TypedTupleMember @type type_name:[_]]");
    const matches = cursor.query([query]);

    let match = null;
    while ((match = matches.next())) {
      const captures = match.captures;
      const cursors = captures["type"];

      const cursor = cursors?.[0];

      names.push((cursor?.node() as NonterminalNode).unparse());
    }

    assert.deepStrictEqual(names, ["uint", " uint16", " uint64", " uint256"]);
    // --8<-- [end:matching-on-label]
  }

  {
    // Matching on node's literal value
    const parse_output = await parseDocInputFile("typed-tuple.sol");
    const cursor = parse_output.createTreeCursor();
    // --8<-- [start:matching-on-literal-value]

    const names = [];

    const query = Query.parse(`[ElementaryType @uint_keyword variant:["uint"]]`);
    const matches = cursor.query([query]);

    let match = null;
    while ((match = matches.next())) {
      const captures = match.captures;
      const cursors = captures["uint_keyword"];

      const cursor = cursors?.[0];

      names.push((cursor?.node() as TerminalNode).text);
    }

    assert.deepStrictEqual(names, ["uint"]);
    // --8<-- [end:matching-on-literal-value]
  }

  {
    const parse_output = await parseDocInputFile("tx-origin.sol");
    const cursor = parse_output.createTreeCursor();
    // --8<-- [start:tx-origin]
    const query = Query.parse(`
    @txorigin [MemberAccessExpression
      [Expression @start ["tx"]]
      [MemberAccess ["origin"]]
    ]`);
    const matches = cursor.query([query]);

    const found = [];

    let match = null;
    while ((match = matches.next())) {
      const captures = match.captures;
      const cursors = captures["txorigin"];

      const cursor = cursors?.[0];

      found.push([cursor?.textOffset.utf8, (cursor?.node() as NonterminalNode).unparse()]);
    }

    assert.deepStrictEqual(found, [[375, "tx.origin"]]);
    // --8<-- [end:tx-origin]
  }
});

File: crates/solidity/outputs/cargo/tests/src/doc_examples/using_queries.rs
use std::path::Path;

use anyhow::Result;
use infra_utils::paths::PathExtensions;
use semver::Version;
use slang_solidity::cst::{NonterminalKind, Query, QueryMatchIterator};
use slang_solidity::parser::{ParseOutput, Parser};

fn parse_doc_input_file<T: AsRef<Path>>(path: T) -> Result<ParseOutput> {
    let input_path = Path::repo_path("documentation/public/user-guide/inputs").join(path.as_ref());

    let source = input_path.read_to_string()?;

    let parser = Parser::create(Version::new(0, 8, 0))?;

    Ok(parser.parse(NonterminalKind::SourceUnit, source.trim()))
}

#[test]
fn using_queries() -> Result<()> {
    #[allow(unused_variables, clippy::items_after_statements)]
    {
        let parse_output = parse_doc_input_file("using-the-cursor.sol")?;
        // --8<-- [start:creating-a-query]
        use slang_solidity::cst::Query;

        // Any `Cursor` can be used to create a query.
        let cursor = parse_output.create_tree_cursor();

        let query = Query::parse("[ContractDefinition]").unwrap();
        let result: QueryMatchIterator = cursor.query(vec![query]);
        // --8<-- [end:creating-a-query]
    }

    {
        let parse_output = parse_doc_input_file("using-the-cursor.sol")?;
        let cursor = parse_output.create_tree_cursor();
        // --8<-- [start:visiting-contracts]
        let mut found = vec![];

        let query = Query::parse("@contract [ContractDefinition]").unwrap();

        for r#match in cursor.query(vec![query]) {
            let captures = r#match.captures;
            let cursors = captures.get("contract").unwrap();

            let cursor = cursors.first().unwrap();

            found.push(cursor.node().unparse().trim().to_owned());
        }

        assert_eq!(
            found,
            ["contract Foo {}", "contract Bar {}", "contract Baz {}"]
        );
        // --8<-- [end:visiting-contracts]
    }

    {
        let parse_output = parse_doc_input_file("multiple-data-types.sol")?;
        let cursor = parse_output.create_tree_cursor();
        // --8<-- [start:multiple-patterns]
        let mut names = vec![];

        let struct_def = Query::parse("[StructDefinition @name [Identifier]]").unwrap();
        let enum_def = Query::parse("[EnumDefinition @name [Identifier]]").unwrap();

        for r#match in cursor.query(vec![struct_def, enum_def]) {
            let index = r#match.query_number;
            let captures = r#match.captures;
            let cursors = captures.get("name").unwrap();

            let cursor = cursors.first().unwrap();

            names.push((index, cursor.node().unparse()));
        }

        assert_eq!(
            names,
            &[
                (0, "Foo".to_string()),
                (1, "Bar".to_string()),
                (0, "Baz".to_string()),
                (1, "Qux".to_string())
            ]
        );
        // --8<-- [end:multiple-patterns]
    }

    {
        let parse_output = parse_doc_input_file("typed-tuple.sol")?;
        let cursor = parse_output.create_tree_cursor();
        // --8<-- [start:matching-on-label]

        let mut names = vec![];

        let query = Query::parse("[TypedTupleMember @type type_name:[_]]").unwrap();

        for r#match in cursor.query(vec![query]) {
            let captures = r#match.captures;
            let cursors = captures.get("type").unwrap();

            let cursor = cursors.first().unwrap();

            names.push(cursor.node().unparse());
        }

        assert_eq!(names, &["uint", " uint16", " uint64", " uint256"]);
        // --8<-- [end:matching-on-label]
    }

    {
        // Matching on node's literal value
        let parse_output = parse_doc_input_file("typed-tuple.sol")?;
        let cursor = parse_output.create_tree_cursor();
        // --8<-- [start:matching-on-literal-value]

        let mut names = vec![];

        let query = Query::parse(r#"[ElementaryType @uint_keyword variant:["uint"]]"#).unwrap();

        for r#match in cursor.query(vec![query]) {
            let captures = r#match.captures;
            let cursors = captures.get("uint_keyword").unwrap();

            let cursor = cursors.first().unwrap();

            names.push(cursor.node().unparse());
        }

        assert_eq!(names, &["uint"]);
        // --8<-- [end:matching-on-literal-value]
    }

    Ok(())
}

#[test]
fn tx_origin_query() -> Result<()> {
    let parse_output = parse_doc_input_file("tx-origin.sol")?;
    let cursor = parse_output.create_tree_cursor();
    // --8<-- [start:tx-origin]
    let query = Query::parse(
        r#"@txorigin [MemberAccessExpression
                [Expression @start ["tx"]]
                ["origin"]
            ]"#,
    )
    .unwrap();

    let mut results = vec![];

    for r#match in cursor.query(vec![query]) {
        let captures = r#match.captures;
        let cursors = captures.get("txorigin").unwrap();

        let cursor = cursors.first().unwrap();

        results.push((cursor.text_offset().utf8, cursor.node().unparse()));
    }

    assert_eq!(results, &[(375usize, "tx.origin".to_string())]);
    // --8<-- [end:tx-origin]

    Ok(())
}
